kuvert_mta_wrapper.man
.Dd October 25, 2001
.Dt KUVERT_MTA_WRAPPER 1
.Os Unix
.Sh NAME
kuvert-mta-wrapper \- wrapper around your MTA for mail submission to 
.Xr kuvert 1
.Sh SYNOPSIS
.Nm kuvert-mta-wrapper
.Op Fl options 
.Op Ar args
.Sh DESCRIPTION
kuvert_mta_wrapper submits an email either directly to your MTA or
enqueues it for 
.Xr kuvert 1
for further processing. kuvert_mta_wrapper should be called by your MUA
instead of your usual MTA in order to enable kuvert to intercept and
process the outgoing mails. Please see your MUA's documentation about
how to override the MTA to be used.
.Pp
The decision whether queueing or calling the MTA directly is based on 
the following factors:
.Bl -enum
.It 
If there are options given other than 
.Op Fl "bm",
.Op Fl "f",
.Op Fl "i",
.Op Fl "t",
.Op Fl "v",
.Op Fl "oi",
.Op Fl "d",
.Op Fl "e",
the standard MTA 
.Pa /usr/lib/sendmail
is executed with the options and arguments given. The result code in this case
is the one the MTA returns.
.It
If there is a configuration file 
.Pa ~/.kuvert
(See 
.Xr kuvert "1" for possible configuration directives) the mail is queued for 
.Xr kuvert "1". The options and arguments are ignored.
If there are problems, an error message is sent to syslog and -1 is returned.
.El
.Sh FILES
.Bl -tag
.It Pa ~/.kuvert
configuration file for 
.Xr kuvert "1" and
.Xr kuvert_mta_wrapper "1".
.It Pa ~/.kuvert_queue
the default queue directory for 
.Xr kuvert "1" if the configuration file does not specify an alternative.
.El
.Sh SEE ALSO
.Xr kuvert 1
.Sh AUTHORS
.An Alexander Zangerl <az@snafu.priv.at>
.Sh BUGS
The list of allowed options and the MTA for fallback are set at compile time.


debian/NEWS
kuvert (2.0.0) unstable; urgency=low

  kuvert has been completely revamped, with a number of new features
  and minus some legacy stuff.
  
  New: 
   full inbound and outbound support for SMTP
   support for gpg-agent
   simplified action settings
   overridable keys and actions in the comments of From: and To:
   simpler submission wrapper program

  Gone: 
   support for pgp2 (but RSA keys continue to work fine, with gpg)
   kuvert no longer stores passphrases itself, ever.
    
  Because of these changes the new config file format is different. 
  The new kuvert will not run with an old-style config file, but 
  does a rough auto-conversion on startup. It'll dump that skeleton
  in /tmp and tell you about it. You will need to go over that and 
  the kuvert manpage and adjust the config to your liking.

 -- Alexander Zangerl <az@debian.org>  Sun, 29 Jun 2008 21:32:14 +1000

kuvert (1.1.10) unstable; urgency=low

  The configuration options AGENTPATH and CLIENTPATH have been
  deprecated and support for a private q-agent process was dropped;
  while kuvert still suggests quintuple-agent, it is no longer favouring
  it and will work with any (reasonable) external passphrase store.
  
  kuvert will not start until you update your 
  personal .kuvert configuration file and remove AGENTPATH/CLIENTPATH
  and use the new GETSECRET and DELSECRET directives if you want
  to use an external passphrase store. 

 -- Alexander Zangerl <az@debian.org>  Fri,  4 Nov 2005 16:43:14 +1000



kuvert_mta_wrapper.c
/*
 * $Id: kuvert_mta_wrapper.c,v 1.6 2003/03/28 10:55:39 az Exp az $
 * 
 * this file is part of kuvert, a wrapper around your mta that
 * does pgp/gpg signing/signing+encrypting transparently, based
 * on the content of your public keyring(s) and your preferences.
 *
 * copyright (c) 1999-2003 Alexander Zangerl <az+kuvert@snafu.priv.at>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <stdio.h>
#include <pwd.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <ctype.h>
#include <syslog.h>

#define CONFFILE "/.kuvert"
#define DEFAULT_QUEUEDIR "/.kuvert_queue"
#define BUFLEN 65536
#define FALLBACKMTA "/usr/lib/sendmail"

#define BAILOUT(a,...) {fprintf(stderr,"%s: ",argv[0]); fprintf(stderr, a "\n",##__VA_ARGS__);syslog(LOG_ERR,a,##__VA_ARGS__); exit(1);}

int main(int argc,char **argv)
{
   struct passwd *pwentry;
   /* fixme sizes */
   char filen[256],buffer[BUFLEN],dirn[256];
   int res,c,fallback,spaceleft;
   char *p,*dirnp;
   FILE *out;
   FILE *cf;
   struct stat statbuf;

   /* determine whether to queue stuff or to call sendmail
      directly: if there is a proper config file of kuvert in $HOME,
      and if the flags/args given are "consistent" with a call
      to sendmail for mail submission, do queue stuff;
      otherwise exec sendmail. */

   openlog(argv[0],LOG_NDELAY|LOG_PID,LOG_MAIL);
   
   /* scan the arguments for options:
      we understand about: no options, non-option-args, --,
      -bm, -f, -i, -t, -v, -m, -oi, -d*, -e*. everything else means some special
      instruction to sendmail, so we exec sendmail. */

   /* no getopt error messages, please! */
   opterr=0;

   while ((c=getopt(argc,argv,"f:itvb:mo:"))!=-1 && !fallback)
   {
      switch (c)
      {
	 case 'v':
	 case 'f':
	 case 'i':
	 case 't':
	 case 'm':			/* deprecated option 'metoo',
					   but nmh uses this... */
	    break;			/* these options are ok and supported */
	 case 'b':
	    /* just -bm is ok, other -b* are bad */
	    if (!optarg || *optarg != 'm') 
	    {
	       fallback=1;
	       syslog(LOG_INFO,"option '-%c%s' mandates fallback",
		      c,optarg ? optarg : "");
	    }
	    break;
	 case 'o':
	    /* -oi, -oe*, -od* are ok */
	    if (!optarg || (*optarg != 'i' && *optarg != 'e' 
			    && *optarg != 'd'))
	    {
	       fallback=1;
	       syslog(LOG_INFO,"option '-%c%s' mandates fallback",
		      c,optarg ? optarg : "");
	    }
	    break;
	 default:
	    /* well, there's an option we do not know, lets bail out */
	    fallback=1;
	    syslog(LOG_INFO,"option '-%c' mandates fallback",
		   c=='?'?optopt:c);
	    break;
      }
   }
   if (!fallback)
   {
      /* options seem ok, look for config file in $HOME */
      pwentry=getpwuid(getuid());
      if (!pwentry)
	 BAILOUT("getpwuid failed: %s",strerror(errno));
    
      /* open and scan the conffile for an queue-file definition 
	 if there is no conffile, kuvert wont work ever  */
      if (snprintf(filen,sizeof(filen),"%s%s",pwentry->pw_dir,CONFFILE)==-1)
	 BAILOUT("overlong filename, suspicious",NULL);
      if (!(cf=fopen(filen,"r")))	
      {
	 /* no config file -> exec sendmail */
	 syslog(LOG_INFO,"user has no .kuvert config file, fallback");
	 fallback=1;
      }
      else
      {
	 /* scan the lines for ^QUEUEDIR\s+ */
	 dirnp=NULL;
	 while(!feof(cf))
	 {
	    p=fgets(buffer,sizeof(buffer)-1,cf);
	    /* empty file? ok, we'll ignore it */
	    if (!p)
	       break;
	
	    if (!strncmp(buffer,"QUEUEDIR",sizeof("QUEUEDIR")-1))
	    {
	       p=buffer+sizeof("QUEUEDIR")-1;
	       for(;*p && isspace(*p);++p)
		  ;
	       if (*p)
	       {
		  dirnp=p;
		  /* strip the newline from the string */
		  for(;*p && *p != '\n';++p)
		     ;
		  if (*p == '\n')
		     *p=0;
		  /* strip eventual trailing whitespace */
		  for(--p;p>dirnp && isspace(*p);--p)
		     *p=0;
	       }
	       /* empty dir? ignore it */
	       if (strlen(dirnp)<2)
		  dirnp=NULL;
	       break;
	    }
	 }
	 fclose(cf);
      }
   }

   /* fallback to sendmail requested? */
   if (fallback)
   {
      /* mangle argv[0], so that it gets recognizeable by sendmail */
      argv[0]=FALLBACKMTA;
      *buffer=0;

      /* bah, c stringhandling is ugly... i just want all args 
	 in one string for a nice syslog line... */
      for(c=0,spaceleft=sizeof(buffer);
	  c<argc;
	  spaceleft-=strlen(argv[c++]))
      {
	 if (spaceleft <= 0) 
	    BAILOUT("overlong command line, suspicious.",NULL);
	 strncat(buffer,argv[c],spaceleft);
	 --spaceleft && c<argc-1 && strcat(buffer," ");
      }
    
      syslog(LOG_INFO,"will exec MTA as '%s'",buffer);
      execv(FALLBACKMTA,argv);
      /* must not reach here */
      BAILOUT("execv FALLBACKMTA failed: %s",strerror(errno));
   }

   /* otherwise queue the stuff for kuvert,
      first check queuedir and create if missing */
   if (!dirnp)
   {
      if(snprintf(dirn,sizeof(dirn),"%s%s",pwentry->pw_dir,DEFAULT_QUEUEDIR)
	 ==-1)
	 BAILOUT("overlong dirname, suspicous.",NULL);
      dirnp=dirn;
   }

   res=stat(dirnp,&statbuf);
   if (res)
   {
      if (errno == ENOENT)
      {
	 /* seems to be missing -> try to create it */
	 if (mkdir(dirnp,0700))
	    BAILOUT("mkdir %s failed: %s\n",dirnp,strerror(errno));
      }
      else 
	 BAILOUT("stat %s failed: %s\n",dirnp,strerror(errno));
   }
   else if (!S_ISDIR(statbuf.st_mode))
   {  
      BAILOUT("%s is not a directory",dirnp);
   }
   else if (statbuf.st_uid != getuid())
   {
      BAILOUT("%s is not owned by you - refusing to run",dirnp);
   }
   else if ((statbuf.st_mode & 0777) != 0700)
   {
      BAILOUT("%s does not have mode 0700 - refusing to run",dirnp);
   }
   umask(066);			/* absolutely no access for group/others... */
    
   /* dir does exist now */
   snprintf(filen,sizeof(filen),"%s/%d",dirnp,getpid());
  
   /* file create and lock */
   if (!(out=fopen(filen,"a")))
   {
      BAILOUT("fopen %s failed: %s\n",filen,strerror(errno));
   }
   if (flock(fileno(out),LOCK_EX))
   {
      BAILOUT("flock failed: %s\n",strerror(errno));
   }

   /* and put the data there */
   do 
   {
      res=fread(buffer,1,BUFLEN,stdin);
      if (!res && ferror(stdin))
	 BAILOUT("fread failure: %s",strerror(errno));
      if (fwrite(buffer,1,res,out)!=res && ferror(out))
	 BAILOUT("fwrite failure: %s",strerror(errno));
   }
   while (res==BUFLEN);
  
   if (fflush(out)==EOF)
      BAILOUT("fflush failed: %s",strerror(errno));
   if (flock(fileno(out),LOCK_UN))
   {
      BAILOUT("flock (unlock) failed: %s",strerror(errno));
   }
   if (fclose(out)==EOF)
      BAILOUT("fclose failed: %s",strerror(errno));
   return 0;
}
